src/Stratis.Bitcoin/P2P/PeerAddressManager.cs

```diff
@@ -0,0 +1,313 @@
+        }
+
+        internal static int GetRandomInteger(int max)
+        {
+            return (int)(RandomUtils.GetUInt32() % (uint)max);
+        }
+
+        internal void DiscoverPeers(Network network, NodeConnectionParameters nodeParameters, int peersToFind)
+        {
+            this.asyncLoop = this.asyncLoopFactory.Run(nameof(this.DiscoverPeersAsync), async token =>
+            {
+                await DiscoverPeersAsync(network, nodeParameters, peersToFind).ConfigureAwait(false);
+            },
+            nodeParameters.ConnectCancellation,
+            TimeSpans.RunOnce);
+        }
+
+        private async Task DiscoverPeersAsync(Network network, NodeConnectionParameters nodeParameters, int peersToFind)
+        {
+            int peersFound = 0;
+
+            while (peersFound < peersToFind)
+            {
+                nodeParameters.ConnectCancellation.ThrowIfCancellationRequested();
+
+                var peersToDiscover = new List<NetworkAddress>();
+                peersToDiscover.AddRange(SelectPeersToConnectTo());
+
+                if (peersToDiscover.Count == 0)
+                {
+                    PopulateTableWithDNSNodes(network, peersToDiscover);
+                    PopulateTableWithHardNodes(network, peersToDiscover);
+                    peersToDiscover = new List<NetworkAddress>(peersToDiscover.OrderBy(a => RandomUtils.GetInt32()));
+                    if (peersToDiscover.Count == 0)
+                        await Task.CompletedTask;
+                }
+
+                try
+                {
+                    var clonedParameters = nodeParameters.Clone();
+
+                    Parallel.ForEach(peersToDiscover, new ParallelOptions()
+                    {
+                        MaxDegreeOfParallelism = 5,
+                        CancellationToken = nodeParameters.ConnectCancellation,
+                    },
+                    peer =>
+                    {
+                        using (var connectTokenSource = CancellationTokenSource.CreateLinkedTokenSource(nodeParameters.ConnectCancellation))
+                        {
+                            connectTokenSource.CancelAfter(TimeSpan.FromSeconds(5));
+
+                            Node node = null;
+
+                            try
+                            {
+                                clonedParameters.ConnectCancellation = connectTokenSource.Token;
+
+                                var addressManagerBehaviour = clonedParameters.TemplateBehaviors.Find<PeerAddressManagerBehaviour>();
+                                clonedParameters.TemplateBehaviors.Clear();
+                                clonedParameters.TemplateBehaviors.Add(addressManagerBehaviour);
+
+                                node = Node.Connect(network, peer.Endpoint, clonedParameters);
+                                node.VersionHandshake(connectTokenSource.Token);
+                                node.MessageReceived += (s, a) =>
```

Comment:
`  node.MessageReceived += (s, a) =>` is never unsubscribed this is a handle leak.

Reply:
You actually don don't this event, you can replace the condition in the loop 
`while (peersFound < peersToFind)` 
with 
`while (this.peers.Count < peersToFind)`

Reply:
Yes that's correct, I've removed the event 👍